### ISSUE TITLE
Suspected podman-remote container-removal completion race in integration tests.

### ISSUE THREAD

**Issue by cevich** [containers/libpod#4021]
/kind bug

**Description**

A number integration tests expect the presence or absence of containers after a `podman-remote` command returns an exit code.  However, because removal can occur asynchronously to exit, it's possible for these tests to race with removal completion.  Depending on the test, this can either result in a flake, or a false-positive result - i.e. a test sees an expected container, which is removed _after_ the presence-check.

**Steps to reproduce the issue:**

     (example code-path)

1. `podman failed to start with --rm should delete the container (test/e2e/start_test.go)` - `Expect(start.ExitCode()).To(Not(Equal(0)))` is successful but remote container is ***not*** yet removed.

2. Test calls `numContainers := podmanTest.NumberOfContainers()` in `test/utils/utils.go`

3. `ps := p.PodmanBase([]string{"ps", "-aq"}, true, false)` finds the not-yet removed container and counts it.

4. The container is removed

5. The test fails in `Expect(numContainers).To(BeZero())`

     (negative example code-path)

1. `podman failed to start without --rm should NOT delete the container` - `Expect(start.ExitCode()).To(Not(Equal(0)))` is successful.

2. Test calls `numContainers := podmanTest.NumberOfContainers()` in `test/utils/utils.go`

3. `ps := p.PodmanBase([]string{"ps", "-aq"}, true, false)` finds the container and counts it.

4. The container is removed because of some other future bug.

5. The test incorrectly passes because `Expect(numContainers).To(Equal(1))` is true, despite container actually being removed.

**Comment by cevich** [comment]
@mheon @baude (repeated question from #3934)

Is there a way to detect whether or not the removal process has completed or not for a given container ID (like some file that should not exist or something like that)? This would be the ideal solution in both (positive and negative) test scenarios, to prevent racing on an existence-check. For example, test/utils/utils.go could delay until completion before calling ps := p.PodmanBase([]string{"ps", "-aq"}, true, false).

**Comment by rhatdan** [comment]
Removal is supposed to clean everything up.

**Comment by mheon** [comment]
The problem may be that removal is only on the backend, and we may not guarantee that the frontend doesn't exit first

**Comment by rhatdan** [comment]
The backend will not respond to the front-end until the container is removed.

**Comment by cevich** [comment]
> The backend will not respond to the front-end until the container is removed.

Then either we're missing something, that statement is incorrect, or I'm misinterpreting https://cirrus-ci.com/task/5760809608675328?command=integration_test#L6994 based on 82ac0d8

To be clear: The podman ps output shows the container ID that was suppose to be removed, and the wait-for-exit thingie had already returned the non-zero exit expected by the test code.

**Comment by cevich** [comment]
here, let me paste a link to the code where that result came from...

**Comment by cevich** [comment]
...here: https://github.com/containers/libpod/blob/master/test/e2e/start_test.go#L103-L114

**Comment by cevich** [comment]
note: My assumption this is true is because we do ***not*** see a failure coming from `Expect(start.ExitCode()).To(Not(Equal(0)))`, rather it comes from `Expect(numContainers).To(BeZero())` - which should not be possible.

**Comment by cevich** [comment]
> cevich: Do you have a reproducer for https://github.com/containers/libpod/issues/4021 ?  I may have a solution, at least I haven't broken anything :)

@jhonce Reproducer is:  Run our automation a billion times and check each run for it to break :cry: 

What's your thinking?

**Comment by cevich** [comment]
Somehow this (or something else, producing identical symptoms) has reproduced with rootless-remote podman:  https://cirrus-ci.com/task/6286729083879424 on F30

```
[+0668s] [It] podman failed to start with --rm should delete the container
[+0668s]   /var/tmp/go/src/github.com/containers/libpod/test/e2e/start_test.go:104
[+0668s] Running: (env: [PODMAN_VARLINK_ADDRESS=unix:/run/user/1310/io.podman-86d10ff42fd6d2e2455f7229538579792c47b41b7527442a4d7a3f4dbd88c521]) /var/tmp/go/src/github.com/containers/libpod/bin/podman-remote create -it --rm docker.io/library/alpine:latest foo
[+0668s] 186ddd4ae36c7edfe7dbeff8ce5c5c31fbf15981f5446ac652166c1560a1672d
[+0668s] output: 186ddd4ae36c7edfe7dbeff8ce5c5c31fbf15981f5446ac652166c1560a1672d
[+0668s] Running: (env: [PODMAN_VARLINK_ADDRESS=unix:/run/user/1310/io.podman-86d10ff42fd6d2e2455f7229538579792c47b41b7527442a4d7a3f4dbd88c521]) /var/tmp/go/src/github.com/containers/libpod/bin/podman-remote start -l
[+0668s] Error: time="2019-10-01T06:20:44-04:00" level=error msg="container_linux.go:346: starting container process caused \"exec: \\\"foo\\\": executable file not found in $PATH\"\n"
[+0668s] container_linux.go:346: starting container process caused "exec: \"foo\": executable file not found in $PATH": OCI runtime command not found error
[+0668s] output: 
[+0668s] Running: (env: [PODMAN_VARLINK_ADDRESS=unix:/run/user/1310/io.podman-86d10ff42fd6d2e2455f7229538579792c47b41b7527442a4d7a3f4dbd88c521]) /var/tmp/go/src/github.com/containers/libpod/bin/podman-remote ps -aq
[+0668s] 186ddd4ae36c
[+0668s] output: 186ddd4ae36c
[+0758s] [AfterEach] Podman start
[+0758s]   /var/tmp/go/src/github.com/containers/libpod/test/e2e/start_test.go:29
[+0758s] Running: (env: [PODMAN_VARLINK_ADDRESS=unix:/run/user/1310/io.podman-86d10ff42fd6d2e2455f7229538579792c47b41b7527442a4d7a3f4dbd88c521]) /var/tmp/go/src/github.com/containers/libpod/bin/podman-remote stop -a --timeout 0
[+0758s] Running: (env: [PODMAN_VARLINK_ADDRESS=unix:/run/user/1310/io.podman-86d10ff42fd6d2e2455f7229538579792c47b41b7527442a4d7a3f4dbd88c521]) /var/tmp/go/src/github.com/containers/libpod/bin/podman-remote pod stop -a -t 0
[+0758s] output: 
[+0758s] Running: (env: [PODMAN_VARLINK_ADDRESS=unix:/run/user/1310/io.podman-86d10ff42fd6d2e2455f7229538579792c47b41b7527442a4d7a3f4dbd88c521]) /var/tmp/go/src/github.com/containers/libpod/bin/podman-remote pod rm -fa
[+0758s] output: 
[+0758s] Running: (env: [PODMAN_VARLINK_ADDRESS=unix:/run/user/1310/io.podman-86d10ff42fd6d2e2455f7229538579792c47b41b7527442a4d7a3f4dbd88c521]) /var/tmp/go/src/github.com/containers/libpod/bin/podman-remote rm -fa
[+0758s] running: pgrep 28414
[+0758s] 
[+0758s] • Failure [94.043 seconds]
[+0758s] Podman start
[+0758s] /var/tmp/go/src/github.com/containers/libpod/test/e2e/start_test.go:12
[+0758s]   podman failed to start with --rm should delete the container [It]
[+0758s]   /var/tmp/go/src/github.com/containers/libpod/test/e2e/start_test.go:104
[+0758s] 
[+0758s]   Timed out after 90.000s.
[+0758s]   Expected
[+0758s]       <int>: 1
[+0758s]   to be zero-valued
[+0758s] 
[+0758s]   /var/tmp/go/src/github.com/containers/libpod/test/e2e/start_test.go:113
```

To be clear, the test should ***not*** be failing because of:
```
[+0668s] Running: (env: [PODMAN_VARLINK_ADDRESS=unix:/run/user/1310/io.podman-86d10ff42fd6d2e2455f7229538579792c47b41b7527442a4d7a3f4dbd88c521]) /var/tmp/go/src/github.com/containers/libpod/bin/podman-remote ps -aq
[+0668s] 186ddd4ae36c
[+0668s] output: 186ddd4ae36c
```

I'm thinking it's inevitable this can happen on the remote-end, and the test should only fail if eventually the ps does not return '0' containers within some timeout.  In other words, treat this as a testing sensitivity problem rather than a podman-remote problem.

**Comment by jwhonce** [comment]
@cevich If this screen snap is after https://github.com/containers/libpod/pull/4121 merged then that failure is testing for the container count to be zero. As you can see it waited 90s for the container to be removed. The actual "start" command is tested for >0 not ==0 is how I'm determining which test is failing.

**Comment by cevich** [comment]
@jwhonce ya that was after 4121 merged :cry: it's failing on line 113, the start failure is expected.  Failing the `eventually()` on 113 means even after 90-seconds the container was still there...wait a sec, I only see one line of output from 'ps -a' (the container id is correct).  Shouldn't there be 90-seconds / polling interval, worth of output lines?

**Comment by cevich** [comment]
Hmmm, default polling interval is 10ms maybe that's a bit fast and something is hanging (remember, this is `podman-remote`).

**Comment by cevich** [comment]
Dangumit!  I'm still able to reproduce this, even in my polling-interval PR:  https://cirrus-ci.com/task/5631043243081728

**Comment by cevich** [comment]
```
...cut...
[+1672s]   Timed out after 90.000s.
[+1672s]   Expected
[+1672s]       <int>: 1
[+1672s]   to be zero-valued
[+1672s] 
[+1672s]   /var/tmp/go/src/github.com/containers/libpod/test/e2e/start_test.go:113
...cut...
```

@jwhonce what the heck is going on here that I'm not seeing?

**Comment by jwhonce** [comment]
@cevich With your change the timeout is still 90s, you just test less often which should save on resources. :)  You could try  `Eventually(podmanTest.NumberOfContainers(), 180, 3).Should(BeZero())` That would double the wait time and test less often.

**Comment by cevich** [comment]
@jwhonce oh I abandoned that change, it didn't make a difference.  _Somehow_ the problem still reproduces, on my PR and I saw it on master last week too :confused: 

One interesting thing to note is, shouldn't `Eventually(podmanTest.NumberOfContainers()...)...` run the `podman-remote ps -a` command multiple times?  I only see in running once (in the test output) based on the number of 'output: ' lines in the log.  Clue?
